scene/2d/navigation_obstacle_2d.cpp

```diff
@@ -303,30 +255,109 @@ bool NavigationObstacle2D::get_carve_navigation_mesh() const {
 	return carve_navigation_mesh;
 }
 
-void NavigationObstacle2D::_update_map(RID p_map) {
-	map_current = p_map;
-	NavigationServer2D::get_singleton()->obstacle_set_map(obstacle, p_map);
-}
-
 void NavigationObstacle2D::_update_position(const Vector2 p_position) {
 	NavigationServer2D::get_singleton()->obstacle_set_position(obstacle, p_position);
+#ifdef DEBUG_ENABLED
+	if (debug_canvas_item.is_valid()) {
+		Transform2D debug_transform = Transform2D(0.0, get_global_position());
+		RenderingServer::get_singleton()->canvas_item_set_transform(debug_canvas_item, debug_transform);
+	}
+#endif // DEBUG_ENABLED
+}
+
+void NavigationObstacle2D::_update_avoidance_space(RID p_avoidance_space) {
+	NavigationServer2D::get_singleton()->obstacle_set_avoidance_space(obstacle, get_avoidance_space());
+}
+
+void NavigationObstacle2D::_obstacle_enter_tree() {
+	if (!is_inside_tree()) {
+		return;
+	}
+
+	previous_transform = get_global_transform();
+
+	// Need to trigger map controlled agent assignment somehow for the fake_agent since obstacles use no callback like regular agents.
+	NavigationServer2D::get_singleton()->obstacle_set_avoidance_enabled(obstacle, avoidance_enabled);
+	NavigationServer2D::get_singleton()->obstacle_set_avoidance_space(obstacle, get_avoidance_space());
+
+	_update_position(get_global_position());
+	set_physics_process_internal(true);
 #ifdef DEBUG_ENABLED
 	queue_redraw();
 #endif // DEBUG_ENABLED
 }
 
+void NavigationObstacle2D::_obstacle_exit_tree() {
+	set_physics_process_internal(false);
+	_update_avoidance_space(RID());
 #ifdef DEBUG_ENABLED
-void NavigationObstacle2D::_update_fake_agent_radius_debug() {
+	_obstacle_debug_free();
+#endif // DEBUG_ENABLED
+}
+
+void NavigationObstacle2D::_obstacle_physics_process() {
+	if (!is_inside_tree()) {
+		return;
+	}
+
+	_update_position(get_global_position());
+
+	if (velocity_submitted) {
+		velocity_submitted = false;
+		// Only update if there is a noticeable change, else the rvo agent preferred velocity stays the same.
+		if (!previous_velocity.is_equal_approx(velocity)) {
+			NavigationServer2D::get_singleton()->obstacle_set_velocity(obstacle, velocity);
+		}
+		previous_velocity = velocity;
+	}
+
+#ifdef DEBUG_ENABLED
+	queue_redraw();
+#endif // DEBUG_ENABLED
+}
+
+#ifdef DEBUG_ENABLED
+void NavigationObstacle2D::_obstacle_debug_update() {
+	bool is_debug_enabled = false;
+
+	if (Engine::get_singleton()->is_editor_hint()) {
+		is_debug_enabled = true;
+	} else if (NavigationServer2D::get_singleton()->get_debug_enabled() && NavigationServer2D::get_singleton()->get_debug_avoidance_enabled()) {
```

Comment:
perhaps extract `NavigationServer2D::get_singleton()` to variable as this is used in many places